pkg/healthmonitor/health_monitor.go

```diff
@@ -0,0 +1,338 @@
+//
+// Interested parties can register a listener for notifications about healthy/unhealthy targets changes via AddListener.
+// TODO: instead of restConfig we could accept transport so that it is reused instead of creating a new connection to targets
+//       reusing the transport has the advantage of using the same connection as other clients
+func New(targetProvider TargetProvider, restConfig *rest.Config, unhealthyProbesThreshold int, healthyProbesThreshold int, probeResponseTimeout, probeInterval time.Duration) (*HealthMonitor, error) {
+	client, err := createHealthCheckHTTPClient(probeResponseTimeout, restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	hm := &HealthMonitor{
+		client:                   client,
+		targetProvider:           targetProvider,
+		targetsToMonitor:         targetProvider.CurrentTargetsList(),
+		probeResponseTimeout:     probeResponseTimeout,
+		probeInterval:            probeInterval,
+		unhealthyProbesThreshold: unhealthyProbesThreshold,
+		healthyProbesThreshold:   healthyProbesThreshold,
+
+		consecutiveSuccessfulProbes: map[string]int{},
+		consecutiveFailedProbes:     map[string][]error{},
+	}
+	hm.exportedHealthyTargets.Store([]string{})
+	hm.exportedUnhealthyTargets.Store([]string{})
+
+	if notifier, ok := targetProvider.(Notifier); ok {
+		notifier.AddListener(hm)
+	}
+
+	return hm, nil
+}
+
+// Run starts monitoring the provided targets until stop channel is closed
+// This method is blocking and it is meant to be launched in a separate goroutine
+func (sm *HealthMonitor) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting the health monitor with Interval = %v, Timeout = %v, HealthyThreshold = %v, UnhealthyThreshold = %v ", sm.probeInterval, sm.probeResponseTimeout, sm.healthyProbesThreshold, sm.unhealthyProbesThreshold)
+	defer klog.Info("Shutting down the health monitor")
+
+	wait.Until(sm.healthCheckRegisteredTargets, sm.probeInterval, stopCh)
+}
+
+// Enqueue schedules refreshing the target list on the next probeInterval
+// This method is used by the TargetProvider to notify that the list has changed
+func (sm *HealthMonitor) Enqueue() {
+	sm.refreshTargetsLock.Lock()
+	defer sm.refreshTargetsLock.Unlock()
+	sm.refreshTargets = true
+}
+
+// Targets returns a list of healthy and unhealthy targets
+func (sm *HealthMonitor) Targets() ([]string, []string) {
+	return sm.exportedHealthyTargets.Load().([]string), sm.exportedUnhealthyTargets.Load().([]string)
+}
+
+// AddListener adds a listener to be notified when the list of healthy targets changes
+//
+// Note:
+// this method is not thread safe and mustn't be called after calling StartMonitoring() method
+func (sm *HealthMonitor) AddListener(listener Listener) {
+	sm.listeners = append(sm.listeners, listener)
+}
+
+type targetErrTuple struct {
+	target string
+	err    error
+}
+
+// refreshTargetsLocked updates the internal targets list to monitor if it was requested (via the Enqueue method)
```

Comment:
will this allow to dynamically add/remove endpoints? i wonder whether we want to support it or not... I would rather lock adding new endpoints

Reply:
yes, we already know that the list can be dynamic or static. For example, for the aggregator the list will be dynamic